scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/remotes/postgres/GoalBuilder.java

```diff
@@ -138,9 +144,49 @@ private static ActivityCreationTemplate makeActivityTemplate(
       }
     }
     builder = builder.ofType(type);
-    for (final var argument : activityTemplate.arguments().entrySet()) {
+    for (final var argument : canonicalizeSerialization(type, activityTemplate.arguments()).entrySet()) {
       builder = builder.withArgument(argument.getKey(), argument.getValue());
     }
     return builder.build();
   }
+
+  /**
+   * Take a set of arguments and run it through getEffectiveArguments to make sure that the SerializedValues
+   * are the correct type. E.g. Int -> Double.
+   *
+   * The set of keys returned will be the same set of keys passed in.
+   */
+  private static Map<String, SerializedValue> canonicalizeSerialization(
+      final ActivityType type,
+      final Map<String, SerializedValue> arguments
+  ) throws GoalBuilderFailure
+  {
+    final var res = new HashMap<String, SerializedValue>();
+    try {
+      final var effectiveArguments = type.getSpecType().getEffectiveArguments(arguments);
+      // If no exception was thrown, we can copy out the subset of the effective arguments that we're interested in.
+      for (final var argument : arguments.entrySet()) {
+        res.put(argument.getKey(), effectiveArguments.get(argument.getKey()));
+      }
+      return res;
+    } catch (InvalidArgumentsException e) {
+      // If an exception is thrown, we still want to extract the valid arguments. Extraneous arguments are left as they were.
+      for (final var argument : e.validArguments) {
+        // This fills in the "correct" serialization of the argument. E.g. Int -> Double
+        res.put(argument.parameterName(), argument.serializedValue());
+      }
+      if (!e.extraneousArguments.isEmpty()) {
+        // If we hit this case, that means the typescript static checking has failed to catch extraneous arguments.
+        throw new GoalBuilderFailure("Activity Template had extraneous arguments: %s".formatted(e.extraneousArguments));
```

Comment:
I'd put this `if (...) throw` before the `for` loop above, that way you don't construct `res` and then just throw :)